plugins/s3_auth/test/aws_auth_v4_test.cc

```diff
@@ -0,0 +1,897 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+/**
+ * @file aws_auth_v4_test.cc
+ * @brief Unit tests for functions implementing S3 auth version 4
+ */
+
+#include <openssl/hmac.h>   /* EVP_MAX_MD_SIZE */
+#include "gtest/gtest.h"    /* Google Test */
+#include "../aws_auth_v4.h" /* S3 auth v4 utility */
+
+/* UriEncode ******************************************************************************************************************* */
+
+TEST(UrlEncode, EncodeEmptyInput)
+{
+  String in("");
+  String encoded = uriEncode(in, /* isObjectName */ false);
+  EXPECT_EQ(0, encoded.length()); /* 0 encoded because of the invalid input */
+}
+
+TEST(UrlEncode, EncodeUnreservedChars)
+{
+  const String in = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                    "abcdefghijklmnopqrstuvwxyz"
+                    "0123456789"
+                    "-._~";
+  String encoded = uriEncode(in, /* isObjectName */ false);
+
+  EXPECT_EQ(in.length(), encoded.length());
+  EXPECT_STREQ("ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+               "abcdefghijklmnopqrstuvwxyz"
+               "0123456789"
+               "-._~",
+               encoded.c_str());
+}
+
+TEST(UrlEncode, EncodeReservedCharsNotObjectName)
+{
+  const String in = " /!\"#$%&'()*+,:;<=>?@[\\]^`{|}"; /* some printable but reserved chars */
+  String encoded  = uriEncode(in, /* isObjectName */ false);
+
+  EXPECT_EQ(3 * in.length(), encoded.length()); /* size of "%NN" = 3 */
+  EXPECT_STREQ("%20%2F%21%22%23%24%25%26%27%28%29%2A%2B%2C%3A%3B%3C%3D%3E%3F%40%5B%5C%5D%5E%60%7B%7C%7D", encoded.c_str());
+}
+
+TEST(UrlEncode, EncodeReservedCharsObjectName)
+{
+  const String in = " /!\"#$%&'()*+,:;<=>?@[\\]^`{|}"; /* some printable but reserved chars */
+  String encoded  = uriEncode(in, /* isObjectName */ true);
+
+  EXPECT_EQ(3 * in.length() - 2, encoded.length()); /* size of "%NN" = 3, '/' is not encoded */
+  EXPECT_STREQ("%20/%21%22%23%24%25%26%27%28%29%2A%2B%2C%3A%3B%3C%3D%3E%3F%40%5B%5C%5D%5E%60%7B%7C%7D", encoded.c_str());
+}
+
+/* Base16Encode **************************************************************************************************************** */
+
+TEST(Base16Encode, Base16EncodeEmpty)
+{
+  const char *in = nullptr;
+  size_t inLen   = 0;
+  String encoded = base16Encode(in, inLen);
+
+  EXPECT_EQ(0, encoded.length());
+}
+
+TEST(Base16Encode, Base16EncodeRFC4648TestVectors)
+{
+  /* use the test vectors from RFC4648: https://tools.ietf.org/html/rfc4648#section-10 (just convert to lower case) */
+  const char *bench[] = {"",       "",     "f",        "66",    "fo",         "666f",   "foo",
+                         "666f6f", "foob", "666f6f62", "fooba", "666f6f6261", "foobar", "666f6f626172"};
+
+  for (int i = 0; i < sizeof(bench) / sizeof(char *); i += 2) {
```

Comment:
Good point, will change it to make all C++11-mongers happy 👍 ... looks better too :)